pandas/core/arrays/base.py

```diff
@@ -488,9 +488,15 @@ def to_numpy(
         """
         result = np.asarray(self, dtype=dtype)
         if copy or na_value is not lib.no_default:
-            result = result.copy()
-        if na_value is not lib.no_default:
-            result[self.isna()] = na_value
+            # We only consider copying if `self` implements a custom numpy
+            # array container, i.e. if the __array__function in implemented. If
+            # not, copying is done anyway by numpy.asarray. Also, numpy.asarray
+            # might copy the __array__() nonetheless if the dtype didn't match,
+            # so we only copy if the result is a reference to the input array.
+            if hasattr(self, "__array__") and result is self.__array__():
```

Comment:
doesn't this mean we'll be calling `self.__array__` twice?

Reply:
That's true! First I thought calling it is cheap anyway, but now that you mention it I realize that it might not be cheap depending on how the array container is implemented. But I can change the logic of this `to_numpy` function to only call it once by checking if the `__array__` interface is implemented before and then having a separate code branch. Okay?

Reply:
Hold on, actually if `__array__` is expensive i.e. because it creates a new array, then this is now yet another case because then we should also not dot the copying later. But I can't think of a way to check for this... Requires some more thought I suppose

Reply:
One solution is to add this method to the ExtenstionArray API that was suggested in the original issue #24345:
```Python
def _to_numpy(self, dtype=None copy=False): -> Union[ndarray, bool]
    # returns ndarray, did_copy
```